Mathlib/Data/ListM/AStar.lean

```diff
@@ -0,0 +1,240 @@
+updating the `explored` and `remaining` fields of `VertexData` at the same time. -/
+partial def nextEdge : M m P V E (VertexData m P V E × E) := do
+  let σ ← get
+  match σ.queue.min with
+  | none => failure
+  | some (p, []) => do
+      -- No nodes left at the lowest priority, remove that priority from the queue and try again.
+      set { σ with queue := σ.queue.erase p }
+      nextEdge
+  | some (p, v :: others) => match σ.data.find? v with
+    | none => failure -- shouldn't be possible: everything in the queue has data
+    | some d => do match ← (uncons (d.remaining) : m _) with
+      | none => do
+          -- No neighbours left to explore, remove this vertex from the queue and try again.
+          set { σ with queue := σ.queue.insert p others }
+          nextEdge
+      | some (e, remaining) => do
+          -- Update the explored and remaining edges at this vertex, and return the edge.
+          let d' := { d with explored := e :: d.explored, remaining := remaining }
+          set { σ with data := σ.data.insert v d' }
+          pure (d', e)
+
+/-- Perform one step of A^* search:
+mark the next edge leaving the highest priority vertex `v` as explored,
+and record the path obtained by adding the edge to the best yet path to `v`. -/
+def update (Γ : GraphData m P V E) (optimal : Bool) :
+    (M m P V E) PUnit := do
+  let (d, e) ← nextEdge
+  recordPath Γ optimal (cons Γ e d.bestPath)
+
+end AStar
+
+variable {m P V E}
+variable [Inhabited P] [Inhabited V] [Inhabited E]
+
+open AStar
+
+/--
+Perform A^* search on a graph, starting from a given vertex.
+
+The graph should be encoded using `AStar.GraphData`.
+
+If the option `optimal` is false,
+then we don't update best paths when a shorter path is found to an already visited vertex.
+If the heuristic is "consistent", this will never happen.
+However if the heuristic is "admissible" but not "consistent",
+`optimal` must be true if you want the shortest path.
+(In practice, the heuristic is often neither consistent nor admissible,
+and you will need to test whether `optimal` is suitable for your requirements.)
+
+Returns a monadic lazy list consisting of triples `(priority, vertex, path)`
+for each vertex we explore an edge from. `path` is the best path so far to that vertex.
+-/
+def aStarSearchPaths (Γ : GraphData m P V E) (optimal : Bool) (v : V) :
+    ListM m (P × V × List E) :=
+let p := Γ.heuristic v
+let d : VertexData m P V E :=
+  { bestPath := { to := v, edges := [], prio := 0 },
+    heuristicPrio := p,
+    totalPrio := p,
+    explored := [],
+    remaining := Γ.nbhd v }
+let σ : State m P V E :=
+  { queue := RBMap.single 0 [v],
+    data := HashMap.empty.insert v d }
+ListM.iterate (do update Γ optimal; readMin)
+  |>.runState' σ |>.map fun d => ⟨d.totalPrio, d.bestPath.to, d.bestPath.edges⟩
+
+/--
+Perform A^* search on a graph,
+starting at a vertex `src` and ending when reaching a vertex satisfying `goal`.
```

Comment:
Please also describe `optimal` in this docstring.

Should `aStarSearch` and `aStarSearchPaths` refer to each other in their respective docstrings?

Reply:
Yes, good idea!

Done.